File: backend-api/services/crypto/vault.py
"""KEK/DEK primitives + per-record AES-256-GCM for the v3.0 vault sync layer.

This module is the pure-crypto foundation for Stage 2 (AWS sync v0). It
does NOT touch the database, FastAPI, or the wire transport. Higher layers
(Stage 2C+) own envelope serialization, base64 boundaries, versioning,
tombstones, and request/response handling.

Two-layer key model (matches `LoomAssist_Roadmap.md` §4):

    password ──scrypt(salt)──▶ KEK ──AES-GCM──▶ wraps DEK (stored on server)
                                                       │
                                                       └─AES-GCM──▶ each record

Password change rewraps the DEK (cheap) rather than re-encrypting every
record (expensive). See `test_password_change_rewrap`.

Parameter choice — VAULT_SCRYPT_N = 2**17 — is deliberately slower than
the `.loombackup` value in services/crypto/backup.py (2**14). Backup KDF
runs once at user request on a desktop; vault KEK is derived under an
attacker's offline guessing budget if the server is ever compromised.
The threat-model trade-off is in §4.

Boundaries this module enforces:
- Inputs and outputs are raw `bytes`. No base64 here.
- Decrypt failures raise `ValueError` (mirrors backup.py's translation
  from `cryptography.exceptions.InvalidTag` so callers don't import the
  cryptography exception hierarchy).
- Every encrypt call generates a fresh nonce via `secrets.token_bytes`.
"""
from __future__ import annotations


VAULT_SCRYPT_N = 2 ** 17
VAULT_SCRYPT_R = 8
VAULT_SCRYPT_P = 1
VAULT_KEY_LENGTH = 32
DEK_LENGTH = 32
NONCE_LENGTH = 12
GCM_TAG_LENGTH = 16
MIN_SALT_LENGTH = 16


def derive_kek(password: str, salt: bytes) -> bytes:
    from cryptography.hazmat.primitives.kdf.scrypt import Scrypt
    if len(salt) < MIN_SALT_LENGTH:
        raise ValueError(f"salt must be at least {MIN_SALT_LENGTH} bytes")
    kdf = Scrypt(
        salt=salt,
        length=VAULT_KEY_LENGTH,
        n=VAULT_SCRYPT_N,
        r=VAULT_SCRYPT_R,
        p=VAULT_SCRYPT_P,
    )
    return kdf.derive(password.encode("utf-8"))


def generate_salt() -> bytes:
    import secrets as _secrets
    return _secrets.token_bytes(32)


def generate_dek() -> bytes:
    import secrets as _secrets
    return _secrets.token_bytes(DEK_LENGTH)


def wrap_dek(dek: bytes, kek: bytes) -> tuple[bytes, bytes]:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    import secrets as _secrets
    if len(dek) != DEK_LENGTH:
        raise ValueError(f"dek must be {DEK_LENGTH} bytes")
    if len(kek) != VAULT_KEY_LENGTH:
        raise ValueError(f"kek must be {VAULT_KEY_LENGTH} bytes")
    nonce = _secrets.token_bytes(NONCE_LENGTH)
    ciphertext = AESGCM(kek).encrypt(nonce, dek, None)
    return ciphertext, nonce


def unwrap_dek(wrapped: bytes, nonce: bytes, kek: bytes) -> bytes:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    from cryptography.exceptions import InvalidTag
    if len(kek) != VAULT_KEY_LENGTH:
        raise ValueError(f"kek must be {VAULT_KEY_LENGTH} bytes")
    try:
        return AESGCM(kek).decrypt(nonce, wrapped, None)
    except InvalidTag:
        raise ValueError("wrong KEK or corrupted wrapped DEK")


def encrypt_record(plaintext: bytes, dek: bytes) -> tuple[bytes, bytes]:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    import secrets as _secrets
    if len(dek) != DEK_LENGTH:
        raise ValueError(f"dek must be {DEK_LENGTH} bytes")
    nonce = _secrets.token_bytes(NONCE_LENGTH)
    ciphertext = AESGCM(dek).encrypt(nonce, plaintext, None)
    return ciphertext, nonce


def decrypt_record(ciphertext: bytes, nonce: bytes, dek: bytes) -> bytes:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    from cryptography.exceptions import InvalidTag
    if len(dek) != DEK_LENGTH:
        raise ValueError(f"dek must be {DEK_LENGTH} bytes")
    try:
        return AESGCM(dek).decrypt(nonce, ciphertext, None)
    except InvalidTag:
        raise ValueError("wrong DEK or corrupted record")

File: ios-app/LoomKit/Sources/LoomKit/Vault/Vault.swift
import Foundation
import CryptoKit
import Clibsodium

/// KEK/DEK primitives + per-record AES-256-GCM. Byte-compatible port of the
/// desktop's backend-api/services/crypto/vault.py:
///
///     password ──scrypt(salt)──▶ KEK ──AES-GCM──▶ wraps DEK (stored on server)
///                                                        │
///                                                        └─AES-GCM──▶ each record
///
/// Wire conventions shared with the desktop:
/// - GCM ciphertext carries the 16-byte tag APPENDED (Python `cryptography`
///   layout). CryptoKit keeps ciphertext/tag separate; we join and split here.
/// - Nonces are 12 random bytes, fresh per encrypt call. No base64 here —
///   raw bytes in, raw bytes out; the sync layer owns encoding.
/// - Decrypt failures throw `VaultError.decryptFailed` (the counterpart of
///   vault.py translating InvalidTag to ValueError).
///
/// scrypt comes from libsodium's `crypto_pwhash_scryptsalsa208sha256_ll` —
/// standard scrypt with explicit N/r/p. CryptoKit has no scrypt; this is the
/// CryptoKit-plus-libsodium split the Stage 3 plan called out.
public enum Vault {
    public static let keyLength = 32
    public static let dekLength = 32
    public static let nonceLength = 12
    public static let gcmTagLength = 16
    public static let minSaltLength = 16

    /// KDF parameters. These travel with the vault (`GET /vault/info` returns
    /// `kdf_params`), so the client honors whatever the vault was created
    /// with rather than assuming the current defaults.
    public struct KDFParams: Codable, Equatable, Sendable {
        public var n: UInt64
        public var r: UInt32
        public var p: UInt32

        public init(n: UInt64 = 1 << 17, r: UInt32 = 8, p: UInt32 = 1) {
            self.n = n
            self.r = r
            self.p = p
        }
    }

    public enum VaultError: Error, Equatable {
        case saltTooShort(minimum: Int)
        case badKeyLength(expected: Int)
        case badNonceLength(expected: Int)
        case kdfFailed
        case decryptFailed
    }

    // MARK: - KDF

    public static func deriveKEK(
        password: String, salt: Data, params: KDFParams = KDFParams()
    ) throws -> Data {
        guard salt.count >= minSaltLength else {
            throw VaultError.saltTooShort(minimum: minSaltLength)
        }
        let passwordBytes = Array(password.utf8)
        var kek = Data(count: keyLength)
        let rc = kek.withUnsafeMutableBytes { kekPtr in
            salt.withUnsafeBytes { saltPtr in
                crypto_pwhash_scryptsalsa208sha256_ll(
                    passwordBytes, passwordBytes.count,
                    saltPtr.bindMemory(to: UInt8.self).baseAddress!, salt.count,
                    params.n, params.r, params.p,
                    kekPtr.bindMemory(to: UInt8.self).baseAddress!, keyLength
                )
            }
        }
        guard rc == 0 else { throw VaultError.kdfFailed }
        return kek
    }

    // MARK: - Random material

    public static func generateSalt() -> Data {
        Data(SymmetricKey(size: .bits256).withUnsafeBytes { Data($0) })
    }

    public static func generateDEK() -> Data {
        Data(SymmetricKey(size: .bits256).withUnsafeBytes { Data($0) })
    }

    // MARK: - DEK wrap

    public static func wrapDEK(_ dek: Data, kek: Data) throws -> (wrapped: Data, nonce: Data) {
        guard dek.count == dekLength else { throw VaultError.badKeyLength(expected: dekLength) }
        return try encrypt(plaintext: dek, key: kek)
    }

    public static func unwrapDEK(wrapped: Data, nonce: Data, kek: Data) throws -> Data {
        try decrypt(ciphertext: wrapped, nonce: nonce, key: kek)
    }

    // MARK: - Records

    public static func encryptRecord(_ plaintext: Data, dek: Data) throws -> (ciphertext: Data, nonce: Data) {
        try encrypt(plaintext: plaintext, key: dek)
    }

    public static func decryptRecord(ciphertext: Data, nonce: Data, dek: Data) throws -> Data {
        try decrypt(ciphertext: ciphertext, nonce: nonce, key: dek)
    }

    // MARK: - AES-256-GCM core (tag appended, desktop layout)

    private static func encrypt(plaintext: Data, key: Data) throws -> (Data, Data) {
        guard key.count == keyLength else { throw VaultError.badKeyLength(expected: keyLength) }
        var nonceBytes = Data(count: nonceLength)
        nonceBytes.withUnsafeMutableBytes { ptr in
            randombytes_buf(ptr.baseAddress!, nonceLength)
        }
        let nonce = try AES.GCM.Nonce(data: nonceBytes)
        let box = try AES.GCM.seal(plaintext, using: SymmetricKey(data: key), nonce: nonce)
        return (box.ciphertext + box.tag, nonceBytes)
    }

    private static func decrypt(ciphertext: Data, nonce: Data, key: Data) throws -> Data {
        guard key.count == keyLength else { throw VaultError.badKeyLength(expected: keyLength) }
        guard nonce.count == nonceLength else { throw VaultError.badNonceLength(expected: nonceLength) }
        guard ciphertext.count >= gcmTagLength else { throw VaultError.decryptFailed }
        let tag = ciphertext.suffix(gcmTagLength)
        let body = ciphertext.prefix(ciphertext.count - gcmTagLength)
        do {
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: nonce), ciphertext: body, tag: tag
            )
            return try AES.GCM.open(box, using: SymmetricKey(data: key))
        } catch {
            throw VaultError.decryptFailed
        }
    }
}
